### PR TITLE
Copy old PlaneData attributes to new PlaneData

### PR DESCRIPTION
**Pull request by psavery**
Because PlaneData objects have immutable HKLs, when something happens that causes the HKLs to change (such as a change in dmin or sgnum) , a new PlaneData object must be instantiated with the new HKLs.

This PR does two main things:
1. Skips creating a new PlaneData object if the HKLs match. So small dmin changes, for instance, won't create a new PlaneData object.
2. Copies over old PlaneData attributes to the new PlaneData, including exclusions.

The exclusions are copied over by creating a mapping between the old HKLs and the new HKLs, and using this to set the new exclusions based on the old ones.

You can see the exclusions persisting here in hexrdgui:

example gif

Notably as well, other attributes such as the tThWidth and tThMax get preserved when a new PlaneData object is created also.

This, along with hexrd/hexrdgui#791, will resolve hexrd/hexrdgui#774

### CODE REVIEW

**Comment by psavery** [comment]
@saransh13 I want to make sure I understand which attributes on the `Material` object changing can change which HKLs come out of `self.unitcell.getHKLs()`. Can you confirm that I'm right about these? ("yes" if it changes which HKLs come out, "no" for does not).

1. dmin: yes
2. sgnum: yes
3. beamEnergy: no
4. latticeParameters: no
5. atominfo: no

Actually, now that I think about it, because 3 and 4 can change the dmin of the different HKLs, they could potentially change which HKLs come out, because they might move some HKLs above/below the dmin. Is that right? If that's the case, all of those attributes change the HKLs except atominfo...

**Comment by saransh13** [comment]
Hey Patrick,

You’re right. All of them except the atominfo changes the hkl list.

**Comment by joelvbernier** [comment]
Slight correction to the above; changes in latticeParameters *can* (as in will generally) change the generated hkls up to dmin.

**Comment by psavery** [comment]
@joelvbernier Sounds good, I'll fix that soon

**Comment by psavery** [comment]
@saransh13 which of those 5 properties affect the structure factor? All of them?

**Comment by saransh13** [comment]
Hey Patrick,

So the beamenergy, lattice parameter and atominfo will affect the structure
factor. Since sgnum affects some of them, it will indirectly change
structure factors too.
It is independent of dmin.

**Comment by psavery** [comment]
@saransh13 Thanks saransh! I have updated the code to make sure the structure factor gets updated for all of these changes.

This PR is now ready.

PS: I removed the comparison of hkls to skip generating new PData, since generating new PData is not that expensive. It simplifies a little bit of the logic.

File: hexrd/material.py [hunk 1]
```diff
@@ -775,9 +785,9 @@ def _set_atominfo(self, v):
         if v.shape[1] == 4:
             self._atominfo = v
         else:
-            print("Improper syntax, array must be n x 4")
+            raise ValueError("Improper syntax, array must be n x 4")
 
-        return
+        self.update_structure_factor()
```
Comment (saransh13) [inline hexrd/material.py:790]: Not related to your commit, but I noticed that the atominfo is not getting updated in the unitcell class, which is what actually calculates the structure factors.
I'll push a fix for this soon.

**Review by saransh13 — COMMENTED** [review]
Looks good to me overall.

File: hexrd/material.py [hunk 2]
```diff
@@ -230,6 +230,7 @@ def _newUnitcell(self):
     def _hkls_changed(self):
         # Call this when something happens that changes the hkls...
         self._newPdata()
+        self.update_structure_factor()
```
[the remark this answers is not in the document]
Comment (saransh13) [reply hexrd/material.py:234]: Great! LGTM.

**Review by joelvbernier — APPROVED** [review]

**Comment by joelvbernier** [comment]
Ah crap -- @saransh13 looks like the more recent changes in #218 caused a conflict.  Looks like an easy resolution; let me or @psavery know if you need help.

**Comment by joelvbernier** [comment]
@saransh13 and @psavery -- I did my best to resolve the merge conflict with this PR; please double-check.  If it looks good, LGTM!

**Comment by joelvbernier** [comment]
@psavery, can you give a quick check as well?

**Review by saransh13 — COMMENTED** [review]
LGTM

**Review by saransh13 — APPROVED** [review]

**Pull request merged**